### PR TITLE
[feat]:[CDS-73741]: adding header support in trigger expression

### PR DESCRIPTION
**Pull request by RajBaviskar**
## Describe your changes

- adding header support in the trigger expression
Screenshot 2023-07-12 at 4 24 00 PM

Bitbucket test

Screenshot 2023-07-17 at 4 24 35 PM

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50415)

### CODE REVIEW

**Comment by RajBaviskar** [comment]
trigger uts

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java [hunk 1]
```diff
@@ -45,6 +46,10 @@ public Object bind() {
                                  "No Metadata present for planExecution :" + ambiance.getPlanExecutionId()));
     Map<String, Object> jsonObject = TriggerHelper.buildJsonObjectFromAmbiance(metadata.getTriggerPayload());
 
+    if (metadata.getTriggerHeader() != null) {
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java:49]: One question, do you know why the line `jsonObject.put(HEADER, triggerPayload.getHeadersMap());` in `TriggerHelper.buildJsonObjectFromAmbiance` is not achieving the same result as this? Maybe we should remove the line from there if its not working.

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java [hunk 2]
```diff
@@ -216,4 +217,20 @@ public static void stampWebhookIdInfo(NGTriggerEntity ngTriggerEntity, String we
...
     }
   }
+
+  public static Map<String, Object> processTriggerHeader(List<HeaderConfig> triggerHeader) {
+    Map<String, Object> headerJsonObject = new HashMap<>();
+    triggerHeader.forEach(header -> {
+      StringBuilder sb = new StringBuilder();
+      if (header.getValues().size() > 1) {
+        header.getValues().forEach(sb::append);
+        String key = header.getKey().replaceAll("-", "");
```
Comment (vcalasansh) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java:227]: Why do we need to replace "-" by "" in the header keys?

**Review by vcalasansh — COMMENTED** [review]
Added some questions/comments

**Comment by RajBaviskar** [comment]
trigger uts

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java (hunk 1 above)
Comment (RajBaviskar) [reply pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java:49]: above line extracts headers map from the payload. Not sure if it's correct. If you see we were only passing the payload part and not the actual headers.

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java (hunk 2 above)
Comment (RajBaviskar) [reply pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java]: Removed. Not applicable anymore.

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java [hunk 3]
```diff
@@ -216,4 +217,18 @@ public static void stampWebhookIdInfo(NGTriggerEntity ngTriggerEntity, String we
       ngTriggerEntity.getTriggerStatus().setWebhookInfo(WebhookInfo.builder().webhookId(webhookId).build());
     }
   }
+
+  public static Map<String, Object> processTriggerHeader(List<HeaderConfig> triggerHeader) {
+    Map<String, Object> headerJsonObject = new HashMap<>();
+    triggerHeader.forEach(header -> {
+      StringBuilder sb = new StringBuilder();
+      if (header.getValues().size() > 1) {
+        header.getValues().forEach(sb::append);
```
Comment (sribalij) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java:226]: If the size is > 1 it'll append without space right? 
Ideally we can eliminate the if and else if that's the case right? (of course with null check on `header.getValues()`)
```
      header.getValues().stream().forEach(val->{
        headerJsonObject.put(header.getKey(), sb.append(val));
      });
```

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java [hunk 4]
```diff
@@ -216,4 +217,18 @@ public static void stampWebhookIdInfo(NGTriggerEntity ngTriggerEntity, String we
       ngTriggerEntity.getTriggerStatus().setWebhookInfo(WebhookInfo.builder().webhookId(webhookId).build());
     }
   }
+
+  public static Map<String, Object> processTriggerHeader(List<HeaderConfig> triggerHeader) {
+    Map<String, Object> headerJsonObject = new HashMap<>();
```
Comment (sribalij) [inline pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java:222]: Checked Gitlab and headers can be like this
```
Date: Thu, 13 Jul 2023 00:04:17 GMT
Content-Type: application/json;charset=utf-8
Content-Length: 125
X-Harness-Trace-Id: f3a962fd2f486fd2e3c4bd7239c0e9b6
Strict-Transport-Security: max-age=15724800; includeSubDomains
Via: 1.1 google
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000
Connection: close
```

File: pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java (hunk 3 above)
Comment (RajBaviskar) [reply pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java]: removed. not applicable anymore.

**Comment by RajBaviskar** [comment]
trigger uts

**Review by vcalasansh — DISMISSED** [review]
:lgtm:

> Reviewed 4 of 8 files at r1, 1 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50415#-:-N_aLastAXmh1gPhVqim:bnfp4nl)* status: 8 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by archit-harness — CHANGES_REQUESTED** [review]
> Reviewed 4 of 8 files at r1, 1 of 4 files at r2, 3 of 3 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50415)* status: 1 change requests. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java` line 223 at r3](https://reviewable.io/reviews/harness/harness-core/50415#-N_bAlgc4aE859vxjiK6:-N_bAlgc4aE859vxjiK7:b-5ediro) ([raw file](https://github.com/harness/harness-core/blob/ca649e7ed2793e187dc4c5a0eb29fb8ff010bb48/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java#L223)):*
> ```Java
>   public static Map<String, Object> processTriggerHeader(List<HeaderConfig> triggerHeader) {
>     Map<String, Object> headerJsonObject = new HashMap<>();
>     triggerHeader.forEach(header -> { headerJsonObject.put(header.getKey(), header.getValues().get(0)); });
> ```

why not to put complete object directly? is it because we are assuming all header values are list?



_Code quote:_
```Java
header.getValues().get(0)
```
___
*[`pipeline-service/modules/orchestration-beans/src/main/java/io/harness/execution/PlanExecutionMetadata.java` line 70 at r3](https://reviewable.io/reviews/harness/harness-core/50415#-N_bB3wP88hzPv6Eg51V:-N_bB3wP88hzPv6Eg51W:b-hl2r5f) ([raw file](https://github.com/harness/harness-core/blob/ca649e7ed2793e187dc4c5a0eb29fb8ff010bb48/pipeline-service/modules/orchestration-beans/src/main/java/io/harness/execution/PlanExecutionMetadata.java#L70)):*
> ```Java
>   private Boolean executionInputConfigured;
>   @Wither private String triggerJsonPayload;
>   @Wither private List<HeaderConfig> triggerHeader;
> ```

Can we instead have a wrapper class TriggerMetadata and have this inside it?

**Review by RajBaviskar — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50415)* status: 1 change requests. 3. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java` line 223 at r3](https://reviewable.io/reviews/harness/harness-core/50415#-N_bAlgc4aE859vxjiK6:-N_bIpCi5VRp4IiUiEOW:b-blojzz) ([raw file](https://github.com/harness/harness-core/blob/ca649e7ed2793e187dc4c5a0eb29fb8ff010bb48/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/helpers/TriggerHelper.java#L223)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

why not to put complete object directly? is it because we are assuming all header values are list?
</blockquote></details>

some can be list, hence kept it like this. 
___
*[`pipeline-service/modules/orchestration-beans/src/main/java/io/harness/execution/PlanExecutionMetadata.java` line 70 at r3](https://reviewable.io/reviews/harness/harness-core/50415#-N_bB3wP88hzPv6Eg51V:-N_bIzWB7wa3zsdIGb8t:b3nm2gw) ([raw file](https://github.com/harness/harness-core/blob/ca649e7ed2793e187dc4c5a0eb29fb8ff010bb48/pipeline-service/modules/orchestration-beans/src/main/java/io/harness/execution/PlanExecutionMetadata.java#L70)):*
<details><summary><i>Previously, archit-harness (Archit Singla) wrote…</i></summary><blockquote>

Can we instead have a wrapper class TriggerMetadata and have this inside it?
</blockquote></details>

hmm, it might be overkill as we know it will not have any other info in the future.

**Comment by arvind-choudhary-h** [comment]
*[`pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java` line 49 at r3](https://reviewable.io/reviews/harness/harness-core/50415#-N_f5pTp15O1DUoxk0MV:-N_f5pTp15O1DUoxk0MW:bqz688w) ([raw file](https://github.com/harness/harness-core/blob/ca649e7ed2793e187dc4c5a0eb29fb8ff010bb48/pipeline-service/modules/ng-triggers/src/main/java/io/harness/ngtriggers/expressions/functors/TriggerFunctor.java#L49)):*
> ```Java
>     Map<String, Object> jsonObject = TriggerHelper.buildJsonObjectFromAmbiance(metadata.getTriggerPayload());
> 
>     if (metadata.getTriggerHeader() != null) {
> ```

nit: EmptyPredicate.isNotEmpty instead of null

**Comment by RajBaviskar** [comment]
trigger uts

**Review by archit-harness — APPROVED** [review]
> Reviewed 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50415)* status: 3 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by vcalasansh — APPROVED** [review]
:lgtm:

> Reviewed 2 of 3 files at r4, 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50415#-:-N_jRpnYBjwkrRkFOAsp:bnfp4nl)* status: 1. unresolved discussions.

**Comment by RajBaviskar** [comment]
trigger review

**Pull request merged**